content/en/blog/_posts/2021-10-05-nsa-cisa-hardening.md

```diff
@@ -0,0 +1,413 @@
+
+### Etcd
+
+As a general thumb of rule, the etcd server should be configured to only trust
+certs assigned to API server. It limits the attack surface and prevents the
+malicious attacker from gaining access to the cluster. It might be beneficial to
+use a separate CA for etcd, as it by default trusts all the certificates issued
+by root CA.
+
+### Kubeconfig Files
+
+In addition to specifying the token and certificates directly, `.kubeconfig`
+supports dynamic retrieval of temporary tokens using auth provider plugins.
+Beware of the possibility of malicious
+shell [code execution](https://banzaicloud.com/blog/kubeconfig-security/) in a
+`kubeconfig` file. Once attackers gain access to the cluster, they can steal ssh
+keys/secrets or more.
+
+### Secrets
+
+[External secrets providers](https://github.com/external-secrets/kubernetes-external-secrets)
+and [csi-secrets-store](https://github.com/kubernetes-sigs/secrets-store-csi-driver)
+can also be used as alternatives to Kubernetes secrets.
+
+## Log Auditing
+
+The report stresses monitoring and alerting based on logs. The key points
+include logging at the host level, application level, and on the cloud. When
+running kubernetes in production, it's important to understand who's
+responsible, and who's accountable, for each layer of logging.
+
+### Kubernetes API auditing
+
+One area that deserves more focus is what exactly should alert or be logged. The
+document outlines a sample policy in *Appendix L: Audit Policy* that logs all
+RequestResponse's including metadata and request / response bodies. While
+helpful for a demo, it may not be practical for production.
+
+Each organization needs to evaluate their
+own [threat model](https://en.wikipedia.org/wiki/Threat_model) and build an
+audit policy that complements or helps troubleshooting incident response. Think
+about how someone would attack your organization and what audit trail could
+identify it. Review more advanced options for tuning audit logs in
+the [official audit logging documentation](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#audit-policy)
+. It's crucial to tune your audit logs to only include events that meet your
+threat model.
+
+<!–– IF https://github.com/kubernetes-sigs/kind/pull/2431 merges... Audit
+logging configurations can also be tested with
+kind ([instructions](https://deploy-preview-2431--k8s-kind.netlify.app/docs/user/auditing/))
+. ––>
+
+### Streaming logs and auditing
+
+Logging is important for threat and anomaly detection. As the document outlines,
+it's a best practice to scan and alert on logs as close to real time as possible
+and to protect logs from tampering if a compromise occurs. It's important to
+reflect on the various levels of logging and identify the critical areas such as
+API endpoints.
+
+Kubernetes API audit logging can stream to a webhook and there's an example in *
+Appendix N: Webhook configuration*. Using a webhook could be a method that
```

Comment:
```suggestion
Kubernetes API audit logging can invoke a webhook to write batches of events;
there's an example in *Appendix N: Webhook configuration*. Using a webhook
could be a method that
```
This is more like batched writes and not true streaming. You can set batch size == 1.

Reply:
Can we also link "Appendix N: Webhook configuration" to https://media.defense.gov/2021/Aug/03/2002820425/-1/-1/1/CTR_KUBERNETES%20HARDENING%20GUIDANCE.PDF#page=58?